common/autoware_auto_perception_rviz_plugin/include/object_detection/predicted_objects_display.hpp

```diff
@@ -96,24 +98,64 @@ class AUTOWARE_AUTO_PERCEPTION_RVIZ_PLUGIN_PUBLIC PredictedObjectsDisplay
     return id_map.at(uuid);
   }
 
+  // createMarkers: create markers from the message with the single working process; activated when
+  // num_threads<=1;
   std::vector<visualization_msgs::msg::Marker::SharedPtr> createMarkers(
     PredictedObjects::ConstSharedPtr msg);
+  // Working thread is detached from the main to control the flow of messages
   void workerThread();
+  // parallelizedCreateMarkerWorkerThread: create markers from the message with multiple workers
+  void parallelizedCreateMarkerWorkerThread(int rank);
 
   void update(float wall_dt, float ros_dt) override;
 
+  std::vector<visualization_msgs::msg::Marker::SharedPtr> tackle_one_object(
+    PredictedObjects::ConstSharedPtr _msg, int index,
+    std::vector<visualization_msgs::msg::Marker::SharedPtr> _markers);
+
+  void push_tmp_markers(std::vector<visualization_msgs::msg::Marker::SharedPtr> marker_ptrs);
+
   std::unordered_map<boost::uuids::uuid, int32_t, boost::hash<boost::uuids::uuid>> id_map;
   // std::unordered_map<boost::uuids::uuid, int32_t> id_map;
   std::list<int32_t> unused_marker_ids;
   int32_t marker_id = 0;
   const int32_t PATH_ID_CONSTANT = 1e3;
 
+  // msg: latest message, passed from the main callback to workThread;
   PredictedObjects::ConstSharedPtr msg;
+
+  // tmp_msg: latest message, maintained inside working threads and consumed by parallelized workers
+  PredictedObjects::ConstSharedPtr tmp_msg;
   bool consumed{false};
+
+  // mutex: mutex for accessing this->msg, handle conflicts between main process and the working
+  // thread;
   std::mutex mutex;
+
+  // tmp_mutex: mutex for accessing this->tmp_markers, handle conflicts between parallelized
+  // workers;
+  std::mutex tmp_marker_mutex;
+
+  // num_threads: number of threads going to be used, can be modifed on the fly in RVIZ parameters;
```

Comment:
```suggestion
  // num_threads: number of threads going to be used, can be modifed on the fly in RViz parameters;
```

http://wiki.ros.org/rviz